File: benchmarks/fibonacci.js
// Fibonacci benchmark

// Calculate fibonacci recursively
function fibonacciRecursive(n) {
  if (n <= 1) return n;
  return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

// Calculate fibonacci with memoization
function fibonacciMemoized(n, memo = {}) {
  if (n in memo) return memo[n];
  if (n <= 1) return n;
  memo[n] = fibonacciMemoized(n - 1, memo) + fibonacciMemoized(n - 2, memo);
  return memo[n];
}

// Calculate fibonacci iteratively
function fibonacciIterative(n) {
  if (n <= 1) return n;
  let a = 0, b = 1;
  for (let i = 2; i <= n; i++) {
    const c = a + b;
    a = b;
    b = c;
  }
  return b;
}

// Run benchmark
const n = 40;
const startTime = Date.now();

// Use iterative approach as recursive would be too slow for large n
const result = fibonacciIterative(n);

const endTime = Date.now();
const duration = endTime - startTime;

console.log(`Fibonacci(${n}) = ${result}`);
console.log(`Duration: ${duration}ms`);

File: dart_benchmarks/fibonacci.dart
// Fibonacci benchmark for Dart
import 'dart:core';
import 'dart:io';

// Regular recursive implementation (inefficient, for comparison)
int fibonacciRecursive(int n) {
  if (n <= 1) return n;
  return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

// Memoized implementation
Map<int, int> memo = {};
int fibonacciMemoized(int n) {
  if (memo.containsKey(n)) return memo[n]!;
  if (n <= 1) return n;
  
  memo[n] = fibonacciMemoized(n - 1) + fibonacciMemoized(n - 2);
  return memo[n]!;
}

// Iterative implementation (most efficient)
int fibonacciIterative(int n) {
  if (n <= 1) return n;
  
  int fib = 1;
  int prevFib = 1;
  
  for (int i = 2; i < n; i++) {
    int temp = fib;
    fib += prevFib;
    prevFib = temp;
  }
  
  return fib;
}

void main() {
  final n = 40; // Same value as JS benchmark
  
  // Use the most efficient algorithm for benchmarking
  final stopwatch = Stopwatch()..start();
  
  // Run the benchmark
  final result = fibonacciMemoized(n);
  
  // Stop timer
  stopwatch.stop();
  final duration = stopwatch.elapsedMilliseconds;
  
  // Output in same format as JS version
  print('Fibonacci($n) = $result');
  print('Duration: ${duration}ms');
}